### DISCUSSION TITLE
Direct LAN Connection / Inverter Crashing

### DISCUSSION THREAD

**Discussion by pgm83** [nathanmarlor/foxess_modbus#377] [Support]
I'd seen some chat on issues/discussions about this not affecting many people but wanted to confirm that I am experiencing the issue since firmware upgrade. I had no issues with this integration on the older firmware (sorry don't remember exactly what that was, but maybe able to find it if necessary - it was def pre 1.56, i think it was 1.50) but foxess updated mine due to a potential issue with my inverter.

I beleive I'm now on the latest for my H1-5.0-E: Inverter Hardware Version : -Master : 1.60 Slave : 1.02 Manager : 1.63.

My inverter crashes now unfortunately :(

With the integration default settings (on the config page they show as poll 10s and max read as 100) the inverter ran between 1000 - 1545 until it hung.

With adjusted poll rate to 30 seconds the inverter ran between 1633 - 1540 until it hung.

I'm happy to do some testing around this if needed, although I'm going away from Thu to Tue next week so can't do it then. I'm going to disable this plug in while I'm away as the inverter seems fine without it.

Worth noting that I have a wifi bridge via Vonets Mini WiFi Router VAR11N-300 as well in case that is relevant. I don't currently have a way to plug an ethernet cable straight into the inverter because of it's location.

If you want any other info please let me know.

Thanks

**Comment by FozzieUK** [comment]
I’m afraid to say the ethernet connection on the H1 inverter is known to be a bit flaky, Foxess actually removed it from firmware 1.54 because it was crashing some inverters, they then re-enabled (I can’t recall but think it was 1.56) and ever since there have been spurious reports of LAN activity causing it to freeze.

Probably not want you want to hear but the most reliable and guaranteed way to fix this is to change over RS485- this operates flawlessly but will require you to purchase a low cost RS485 interface and then wire 2 cores into your ‘meter’ connector.

But a simple test we can try for now is to reduce the load on the ethernet, you’ve already slowed the polling, can you decrease the max reads to 10, and keep your poll rate at 30 seconds - lets see if that improves things.

**Reply by pgm83** [reply]
It was disabled in 1.56 (I think) and returned in 1.57! Worked fine in 1.50 sadly.

I'm aware of the options around RS485 but I'd need to get power nearby or find something I can power from the inverters usb port, and potentially get an ethernet cable up there too. The Elfin adapater looks promising if I can get it powered via an old usb A cable.

I wasn't sure if the guys would want to do some testing on the Direct LAN connection while they were supporting it, and had noted on other discussions about the Direct LAN/Delay settings that they didn't have any reports of issues and had though perhaps because they could read all registers in one read rather than queing loads up it might not be an issue?

I'm keen to move to RS485 when I can, but in the meantime if I can help test settings for direct lan I'm happy to try.

The reason I hadn't changed the max reads down so far, as that seemed to then cause more reads/potentially queing where the delay:1 required on stealthchestnut might be an issue? I'll do that now in the meantime though and see what happens.

If there are any other combinations that are worth testing let me know.

**Reply by canton7** [reply]
> where the delay:1 required on stealthchestnut might be an issue?

`delay: 1` is a delay which is added after the TCP connection is established but before any data is sent. It does not apply between commands (and doesn't appear to apply on reconnection, but that might be an oversight). This integration automatically does the same thing: we delay for a second after the initial connection when using a direct LAN connection.

**Reply by pgm83** [reply]
Thanks for confirming re delay: 1. I had spotted something that looked similar but wasn't sure if the line:

"self._poll_delay = 30 / 1000 if protocol == SERIAL or adapter.connection_type == LAN else 0" was equivelent or not.

Let me know if the reconnection delay needs testing.

**Reply by FozzieUK** [reply]
@pgm83 The LAN version reads approx 30 registers on each poll, reducing the max reads to 10 will take it longer but it's possible (admittedly a slim possible) that reading a smaller number of registers will reduce the instantaneous workload on the inverter to avoid the freeze, more than 10 reads certainly crashes a USR W610 adaptor.

We've never got to the bottom of the 'why' some inverters work perfectly on LAN (all mine have) and some others have issues - RS485 is the inverters native communications mode for power meters and always works reliably.

The USB outlet on the inverter can be used to supply a low power device so it is possible to power an Elfin through it if you make up a cable yourself, but don't plug anything intelligent into it or the inverter things you are trying to upload firmware and that definitely freezes it - i've even seen a USB power hub knock it out.

**Reply by calum-mcfarlane** [reply]
On my inverter (admittedly on an earlier firmware than I'm running now) I twice saw the inverter 'hardlock' when I'd done nothing more than unplugged and re(?)-plugged in the ethernet cable.

Once I'd made the link between the two events, that was an instant "OK go buy an RS485 interface" moment for me. I don't want the device I'm relying on to power my home to be liable to crashes (with no way to perform a remote restart).

**Reply by canton7** [reply]
> "self._poll_delay = 30 / 1000 if protocol == SERIAL or adapter.connection_type == LAN else 0" was equivelent or not.

That's something else, which adds a 30ms delay after every command, [see here](https://github.com/home-assistant/core/blob/234715a8c68d843bf2f4198e90dfe1728fb77678/homeassistant/components/modbus/modbus.py#L426). The modbus integration adds it to all serial connections, and I additionally added it to LAN connections in case it helped with the crashing thing.

The `delay` logic [taken from here](https://github.com/home-assistant/core/blob/234715a8c68d843bf2f4198e90dfe1728fb77678/homeassistant/components/modbus/modbus.py#L344) is [implemented here](https://github.com/nathanmarlor/foxess_modbus/blob/main/custom_components/foxess_modbus/modbus_client.py#L56).

**Reply by canton7** [reply]
> The LAN version reads approx 30 registers on each poll

We should be [reading 100 per poll by default](https://github.com/nathanmarlor/foxess_modbus/blob/f8719c4d2ca1fdbad9d5770b8d4e18bef7513ced/custom_components/foxess_modbus/inverter_adapters.py#L135)

**Reply by pgm83** [reply]
Inverter ran from 1540 to 0438 with the max read at 10. A very basic non-scientific test I know, but it seems like it's to do with number of connections rather than size? Not sure the inverter is handling the connections well, maybe some kind of memory management issue on the inverter?

**Reply by canton7** [reply]
I'm not sure I follow. There's only ever 1 TCP connection?

**Reply by pgm83** [reply]
Sorry I maybe misunderstood you regarding reconnections mentioned earlier. I guess I also assumed that after each poll it disconnected. 

The time to crash appears that it may be related to the number of reads as it reduced when it had to do more reads? I got circa 24 hours where it works when polling every 30 seconds and reading all registers at once (max 100). When this was changed to max of 10 it ran for around 13 hours before crashing. 

Could it just be that the inverter doesn't manage keeping a tcp connection open for so long and fails to manage some part of it well causing some resource to run out (memory?). Could it be worth trying closing the tcp connection and re-opening a new one after x amount of time or reads?

**Reply by canton7** [reply]
> Sorry I maybe misunderstood you regarding reconnections mentioned earlier. I guess I also assumed that after each poll it disconnected.

No. We'll normally lose the connection only if there's a network problem or similar.

> Could it be worth trying closing the tcp connection and re-opening a new one after x amount of time or reads?

Possibly? It's worth a shot. If I give you a modified file are you happy placing that in the right place on the filesystem?

**Comment by pgm83** [comment]
Yes, happy to try anything as I can't get moved to RS485 connection in the
short term!

**Reply by canton7** [reply]
Cool, give this a try, replacing your existing `modbus_client.py` (probably in `/config/hass/custom_components/foxess_modbus`): https://github.com/nathanmarlor/foxess_modbus/blob/bugfix/lan-disconnection-test/custom_components/foxess_modbus/modbus_client.py . That will drop the connection every minute.

**Comment by pgm83** [comment]
I've uploaded that now, and restarted HA. Not sure if restart was required
or not but it seemed like a sensible thing to do!

I'll leave polling on 30s and max reads on default so we can compare.

**Reply by canton7** [reply]
Thanks! Yep HA will need a restart, should have mentioned that. It might be more interesting to have a poll interval of 10s - that gives us 6 polls between disconnections rather than 2.

**Comment by pgm83** [comment]
ok I've changed that over to default for both poll and read. The plus side
to this is it should crash much quicker if it's going to!

**Comment by pgm83** [comment]
Sadly inverter still crashed in around the same time frame so disconnecting and reconnecting every 60 seconds didn't help. I had hoped that there was some sort of memory bug and maybe the inverter would handle things better if the connection was closed a a new one opened.

I'm not sure if there is much else we can really test here?

**Reply by pgm83** [reply]
That's me back home now. Is there anything else that we can test?

**Reply by FozzieUK** [reply]
I think you’re right, it does sounds like there’s nothing much that will improve things. I think it’s time to plan that RS485 upgrade - if you are struggling for local power use the Elfin EW11a and make (or buy from amazon) an open USB power cable, the Elfin has a wide operating voltage range and can be powered off the inverter’s own USB port.

**Reply by pgm83** [reply]
Thanks Dave.

I'd picked the Elfin out as the best bet for my setup given the location
etc.
Working in the tight space trying to remove the CT connector from the
inverter (assume it screws off?) and then trying to open it without the
tool as I've not been left with it. Add the fact I've an installer who is
awol, if it goes wrong I'm assuming I'm offline for an extended period of
time? not actually sure what happens without the CT clamp, but I'd assume
there would be no way for the inverter to know what load needed matching,
so no battery usage at the very minimum? I'm guessing depending on settings
the PV might be unhappy as well though?

**Reply by FozzieUK** [reply]
It is a little awkward the first time you do it, but if you go carefully it's straight forward.
The connector actually clips onto the inverter, there are 2 serated clips on each side - you push them in and pull the connector straight down from the inverter. - I took some pictures of the clip assembly, and you can see them here along with instructions for connecting the cables https://github.com/nathanmarlor/foxess_modbus/wiki/Modbus-Wiring-Guide

Once the connector is clear of the inverter it's simpler to work with. It can be separated without the tool and using 2 flat bladed screwdrivers, but as the connector is water proof it has a rubber seal which takes a bit of pulling to make the top and bottom separate - it will really help if you have a good look at the photo's so you can work out how it all goes together.

The inverter won't run without the CT connected so no batteries or PV power once it is removed. It is best to switch it off-line and then proceed, as you say the CT wires are only thin multi cores and if the terminal screws in the connector are over-tightened it doesn't take much for them to break - you might need to re-terminate them but it's easy enough to do that - just have all the pin numbers handy. Once you have it all connected any back in place you can restart the inverter.

Where are you located?, if you are around the North West area, i'm happy to pop over (i'll bring the tool) and could give you a hand with it,

**Reply by pgm83** [reply]
Thanks for the guidance, I wasn't too sure on the removal of the connector so good to know it's no tool is required for that part. I hate footery things so attempting to get the connector open with screwdrivers in a tight space in the loft will be fun!

Also thanks for the offer to assist, I am in the North West but probably more North than you'd like, I'm in Glasgow!

I've ordered up the Hyuduo Rs485 To Wifi Serial Server branded version as it was avaialble on amazon prime, so I can attempt this tomorrow if I get the chance!

Randomly, my inverter hasn't crashed after 24 hours this time, still going after 31/32 hours, but I did power down the wifi adapter which in turn disconnected the connection to the inverter 10 seconds at 4pm so it might have something to do with that. Much easier to do that once a day than reboot the entire system after a crash. I wonder it the memory leak whatever is occuring on the inverter may clear when a physical disconnection takes place? I also powered moved the power source of the adapter from the USB port to a mains plug, so that also could play a part I guess. It seems unlikely given the fact we could crash the inverter quicker by changing the polling rate. It just feels like the connection is poorly managed on the inverter side and perhaps uses all available memory after x (circa 2880) no polls.

**Reply by canton7** [reply]
Note that RS485 is not modbus: modbus is another layer on top of RS484. We can support plain serial adapters but you'll find you'll struggle to connect two computers (or two integrations) to it at once, and foxess_em does (yet) support them.

**Reply by canton7** [reply]
> I wonder it the memory leak whatever is occuring on the inverter may clear when a physical disconnection takes place? 

It could be a memory leak, or a race condition, or something overflowing, or going out of bounds, or any number of other things.

**Reply by pgm83** [reply]
Sorry Anthony, I'm not sure what point you are trying to make. Was this in reference to something in my post above? (this was a reply to the RS485/modbus comment!)

**Reply by canton7** [reply]
My bad, edited

**Reply by pgm83** [reply]
Of course. There does seem to be some correlation to no of polls though.

**Reply by pgm83** [reply]
> Note that RS485 is not modbus: modbus is another layer on top of RS484. We can support plain serial adapters but you'll find you'll struggle to connect two computers (or two integrations) to it at once, and foxess_em does (yet) support them.

My bad also, I should have quoted this as it would have made it clearer!

I was wondering what promted the explanation that modbus is seperate to RS485, and the two computers/integrations comment?

**Reply by canton7** [reply]
Sure, all of the things I mentioned could be related to number of polls. We've just no way of knowing - anything is pure unfounded speculation

**Comment by pgm83** [comment]
> Note that RS485 is not modbus: modbus is another layer on top of RS484. We can support plain serial adapters but you'll find you'll struggle to connect two computers (or two integrations) to it at once, and foxess_em does (yet) support them.

My bad also, I should have quoted this as it would have made it clearer!

I was wondering what promted the explanation that modbus is seperate to RS485, and the two computers/integrations comment?

**Reply by canton7** [reply]
The fact that you said you were buying an RS485 to WiFi adapter, instead of the recommended modbus to WiFi adapter.

**Reply by pgm83** [reply]
I see. This was listed on the Wiki here as "The Elfin EW11/EW11A RS485 Modbus to WiFi connector is neat, cheap RS485 to WiFi converter. so I took that to be supported/ok to use?

**Reply by canton7** [reply]
The wiki wording could probably do with a tweak. The one you ordered *should* work, with the caveats I mentioned above.

**Reply by pgm83** [reply]
It's not too late for me to cancel the order if it's not the best one to choose. Which of the devices are true modbus to wifi adapters that would not suffer with the caveats? It makes sense to plan ahead!

**Reply by canton7** [reply]
The wiki lists everything which has been tested (there's one RS485 adapter on there which has suitable caveats)

**Reply by pgm83** [reply]
Thanks.

I guess I'll stick as the only other option on there was the USR610 which seems to have lots of issues and was noted as not recommended.
